### ISSUE TITLE
invalid memory address or nil pointer dereference

### ISSUE THREAD

**Issue by kichik** [wearemolecule/route53-kubernetes#4]
I've been trying to run this as a Kubernetes service for a while now and it keeps failing with the following exception. This happens with both v1.1.0 and v1.1.1.

The node running the container has `elasticloadbalancing:DescribeLoadBalancers` so I don't think it's permissions related. I was able to run that command manually on the same node with Python.

```
I0519 01:14:48.143918       1 service_listener.go:26] Route53 Update Service
I0519 01:14:48.145060       1 service_listener.go:64] Connected to kubernetes @ https://10.3.0.1:443
I0519 01:14:48.145168       1 service_listener.go:93] Starting Service Polling every 30s
I0519 01:14:48.243432       1 service_listener.go:100] Found 1 DNS services in all namespaces with selector "dns=route53"
I0519 01:14:48.243521       1 service_listener.go:115] Creating DNS for nginx service: XXXXX-XXXX.us-XXX.elb.amazonaws.com -> XXXX.YYYY.com
panic: runtime error: invalid memory address or nil pointer dereference
[signal 0xb code=0x1 addr=0x0 pc=0x708590]

goroutine 1 [running]:
github.com/wearemolecule/route53-kubernetes/vendor/github.com/aws/aws-sdk-go/aws/ec2metadata.(*EC2Metadata).GetMetadata(0x0, 0x14800c0, 0x19, 0x0, 0x0, 0x0, 0x0)
        /Users/iterion/Development/go/src/github.com/wearemolecule/route53-kubernetes/vendor/github.com/aws/aws-sdk-go/aws/ec2metadata/api.go:18 +0x1b0
github.com/wearemolecule/route53-kubernetes/vendor/github.com/aws/aws-sdk-go/aws/credentials/ec2rolecreds.requestCredList(0x0, 0x0, 0x0, 0x0, 0x0, 0x0)
        /Users/iterion/Development/go/src/github.com/wearemolecule/route53-kubernetes/vendor/github.com/aws/aws-sdk-go/aws/credentials/ec2rolecreds/ec2_role_provider.go:133 +0x74
github.com/wearemolecule/route53-kubernetes/vendor/github.com/aws/aws-sdk-go/aws/credentials/ec2rolecreds.(*EC2RoleProvider).Retrieve(0xc820212cc0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
        /Users/iterion/Development/go/src/github.com/wearemolecule/route53-kubernetes/vendor/github.com/aws/aws-sdk-go/aws/credentials/ec2rolecreds/ec2_role_provider.go:89 +0x8b
github.com/wearemolecule/route53-kubernetes/vendor/github.com/aws/aws-sdk-go/aws/credentials.(*ChainProvider).Retrieve(0xc820212cf0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
        /Users/iterion/Development/go/src/github.com/wearemolecule/route53-kubernetes/vendor/github.com/aws/aws-sdk-go/aws/credentials/chain_provider.go:75 +0x122
github.com/wearemolecule/route53-kubernetes/vendor/github.com/aws/aws-sdk-go/aws/credentials.(*Credentials).Get(0xc82020ecc0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
        /Users/iterion/Development/go/src/github.com/wearemolecule/route53-kubernetes/vendor/github.com/aws/aws-sdk-go/aws/credentials/credentials.go:185 +0x121
github.com/wearemolecule/route53-kubernetes/vendor/github.com/aws/aws-sdk-go/private/signer/v4.(*signer).sign(0xc820225118, 0x0, 0x0)
        /Users/iterion/Development/go/src/github.com/wearemolecule/route53-kubernetes/vendor/github.com/aws/aws-sdk-go/private/signer/v4/v4.go:182 +0x3c5
github.com/wearemolecule/route53-kubernetes/vendor/github.com/aws/aws-sdk-go/private/signer/v4.Sign(0xc82000db80)
        /Users/iterion/Development/go/src/github.com/wearemolecule/route53-kubernetes/vendor/github.com/aws/aws-sdk-go/private/signer/v4/v4.go:155 +0x1fa
github.com/wearemolecule/route53-kubernetes/vendor/github.com/aws/aws-sdk-go/aws/request.(*HandlerList).Run(0xc82000dca8, 0xc82000db80)
        /Users/iterion/Development/go/src/github.com/wearemolecule/route53-kubernetes/vendor/github.com/aws/aws-sdk-go/aws/request/handlers.go:115 +0x9f
github.com/wearemolecule/route53-kubernetes/vendor/github.com/aws/aws-sdk-go/aws/request.(*Request).Sign(0xc82000db80, 0x0, 0x0)
        /Users/iterion/Development/go/src/github.com/wearemolecule/route53-kubernetes/vendor/github.com/aws/aws-sdk-go/aws/request/request.go:212 +0xce
github.com/wearemolecule/route53-kubernetes/vendor/github.com/aws/aws-sdk-go/aws/request.(*Request).Send(0xc82000db80, 0x0, 0x0)
        /Users/iterion/Development/go/src/github.com/wearemolecule/route53-kubernetes/vendor/github.com/aws/aws-sdk-go/aws/request/request.go:222 +0x54
github.com/wearemolecule/route53-kubernetes/vendor/github.com/aws/aws-sdk-go/service/elb.(*ELB).DescribeLoadBalancers(0xc820024270, 0xc8202e9170, 0xc8202e5400, 0x0, 0x0)
        /Users/iterion/Development/go/src/github.com/wearemolecule/route53-kubernetes/vendor/github.com/aws/aws-sdk-go/service/elb/api.go:608 +0x4f
main.hostedZoneId(0xc820024270, 0xc8202ddb80, 0x47, 0x0, 0x0, 0x0, 0x0)
        /Users/iterion/Development/go/src/github.com/wearemolecule/route53-kubernetes/service_listener.go:178 +0x1ab
main.main()
        /Users/iterion/Development/go/src/github.com/wearemolecule/route53-kubernetes/service_listener.go:121 +0x1f68

goroutine 5 [chan receive]:
github.com/wearemolecule/route53-kubernetes/vendor/github.com/golang/glog.(*loggingT).flushDaemon(0x1b2f360)
        /Users/iterion/Development/go/src/github.com/wearemolecule/route53-kubernetes/vendor/github.com/golang/glog/glog.go:879 +0x67
created by github.com/wearemolecule/route53-kubernetes/vendor/github.com/golang/glog.init.1
        /Users/iterion/Development/go/src/github.com/wearemolecule/route53-kubernetes/vendor/github.com/golang/glog/glog.go:410 +0x297

goroutine 17 [IO wait]:
net.runtime_pollWait(0x7f0c43192a28, 0x72, 0xc82000a130)
        /usr/local/Cellar/go/1.5/libexec/src/runtime/netpoll.go:157 +0x60
net.(*pollDesc).Wait(0xc82020b870, 0x72, 0x0, 0x0)
        /usr/local/Cellar/go/1.5/libexec/src/net/fd_poll_runtime.go:73 +0x3a
net.(*pollDesc).WaitRead(0xc82020b870, 0x0, 0x0)
        /usr/local/Cellar/go/1.5/libexec/src/net/fd_poll_runtime.go:78 +0x36
net.(*netFD).Read(0xc82020b810, 0xc8201e4c00, 0x400, 0x400, 0x0, 0x7f0c43186050, 0xc82000a130)
        /usr/local/Cellar/go/1.5/libexec/src/net/fd_unix.go:232 +0x23a
net.(*conn).Read(0xc8200242a0, 0xc8201e4c00, 0x400, 0x400, 0x0, 0x0, 0x0)
        /usr/local/Cellar/go/1.5/libexec/src/net/net.go:172 +0xe4
crypto/tls.(*block).readFromUntil(0xc820213650, 0x7f0c43192ae8, 0xc8200242a0, 0x5, 0x0, 0x0)
        /usr/local/Cellar/go/1.5/libexec/src/crypto/tls/conn.go:455 +0xcc
crypto/tls.(*Conn).readRecord(0xc82000d8c0, 0x15fe417, 0x0, 0x0)
        /usr/local/Cellar/go/1.5/libexec/src/crypto/tls/conn.go:540 +0x2d1
crypto/tls.(*Conn).Read(0xc82000d8c0, 0xc8201d7000, 0x1000, 0x1000, 0x0, 0x0, 0x0)
        /usr/local/Cellar/go/1.5/libexec/src/crypto/tls/conn.go:901 +0x167
net/http.noteEOFReader.Read(0x7f0c43192fd8, 0xc82000d8c0, 0xc8201f1918, 0xc8201d7000, 0x1000, 0x1000, 0x0, 0x0, 0x0)
        /usr/local/Cellar/go/1.5/libexec/src/net/http/transport.go:1370 +0x67
net/http.(*noteEOFReader).Read(0xc8202e6d20, 0xc8201d7000, 0x1000, 0x1000, 0x0, 0x0, 0x0)
        <autogenerated>:126 +0xd0
bufio.(*Reader).fill(0xc8202e2720)
        /usr/local/Cellar/go/1.5/libexec/src/bufio/bufio.go:97 +0x1e9
bufio.(*Reader).Peek(0xc8202e2720, 0x1, 0x0, 0x0, 0x0, 0x0, 0x0)
        /usr/local/Cellar/go/1.5/libexec/src/bufio/bufio.go:132 +0xcc
net/http.(*persistConn).readLoop(0xc8201f18c0)
        /usr/local/Cellar/go/1.5/libexec/src/net/http/transport.go:876 +0xf7
created by net/http.(*Transport).dialConn
        /usr/local/Cellar/go/1.5/libexec/src/net/http/transport.go:685 +0xc78

goroutine 18 [select]:
net/http.(*persistConn).writeLoop(0xc8201f18c0)
        /usr/local/Cellar/go/1.5/libexec/src/net/http/transport.go:1009 +0x40c
created by net/http.(*Transport).dialConn
        /usr/local/Cellar/go/1.5/libexec/src/net/http/transport.go:686 +0xc9d
```

As it seems to be related to the AWS SDK, I tried updating the dependencies and rebuilding, but that fails because of https://github.com/kubernetes/kubernetes/issues/25823. Updating to latest Kubernetes errors when building on:

```
./service_listener.go:55: undefined: "route53-kubernetes/vendor/k8s.io/kubernetes/pkg/client/unversioned".Config
```

**Comment by iterion** [comment]
It seems like the actual panic is happening in the aws lib: `github.com/aws/aws-sdk-go/aws/ec2metadata.(*EC2Metadata).GetMetadata`.

It seems like a weird place for a kernel panic, but I might be looking at the wrong version on github: "https://github.com/aws/aws-sdk-go/blob/master/aws/ec2metadata/api.go#L18"

I'll try to investigate further, but it does seem as if it could be an issue finding credentials. You could try mounting a secret that has some AWS creds that you know will work:

I do the following snippet in my rc:

``` yaml
          volumeMounts:
            - name: ssl-cert
              mountPath: /opt/certs
              readOnly: true
            - name: aws-creds
              mountPath: /opt/creds
              readOnly: true
          env:
            - name: "CA_FILE_PATH"
              value: "/opt/certs/ca.pem"
            - name: "CERT_FILE_PATH"
              value: "/opt/certs/cert.pem"
            - name: "KEY_FILE_PATH"
              value: "/opt/certs/key.pem"
            - name: "AWS_SHARED_CREDENTIALS_FILE"
              value: "/opt/creds/credentials"
```

Where aws-creds is an ini file that looks like this:

``` ini
[default]
aws_access_key_id = [key]
aws_secret_access_key = [secret]
```

**Comment by kichik** [comment]
That does work. Thanks.
It would still be nice if there was an option to use the EC2 instance role and k8s token so no credentials would have to be directly shared with the pod.

**Comment by iterion** [comment]
Agree, it seems as if there is some bug preventing the pod from getting out to the AWS metadata service. When I get a chance I'll try to investigate that and report back here. Thanks for reporting the issue!

**Comment by kichik** [comment]
Thanks. Let me know if you need any help testing.
Also, on the same topic, it'd be nice if the k8s service token can be used automatically. This actually used to work in v1.1.0 but in v1.1.1 it requires the certificate paths to be in the environment.

**Comment by alph486** [comment]
@iterion I also have run into this issue. I haven't been able to build my fork of this project yet due to transitive dependency issues, but I suspect this has to do with when the `EC2RoleProvider` is invoked when you do not specify a credentials file.

The solution has to do with providing a session to the role as described here: https://github.com/aws/aws-sdk-go/issues/430#issuecomment-155984351 . If I can get it building I'll PR, but if you get a chance @iterion, might be worth a shot.

**Comment by iterion** [comment]
I believe this has been resolved by #9. If anybody still sees this issue as of v1.1.3 please let us know.
